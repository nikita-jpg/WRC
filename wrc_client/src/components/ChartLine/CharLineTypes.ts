export type CharLineDate = {
    dateTime: Date;
    value: number;
};

export type CharLineProps = {
    data: {
        label: string;
        data: CharLineDate[];
        borderColor?: string;
        backgroundColor?: string;
    }[];
    className?: string;
    title?: string;
};
