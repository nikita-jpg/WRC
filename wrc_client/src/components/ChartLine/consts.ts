const DNATranscriber = {
    // withUnitsOfMeasurement
    voltage_withUnitsOfMeasurement: "Напряжение (Вольт)",
    amperage_withUnitsOfMeasurement: "Сила тока (Ампер)",
    power_withUnitsOfMeasurement: "Мощность (Ватт)",
    RPower_withUnitsOfMeasurement:
        "Реактивная мощность (Вольт-амперы реактивные)",
    FPower_withUnitsOfMeasurement: "Полная мощность (Вольт-амперы)",
    "voltage max_withUnitsOfMeasurement": "Максимальное напряжение (Вольт)",
    "voltage min_withUnitsOfMeasurement": "Минимальное напряжение (Вольт)",
    cosinusF_withUnitsOfMeasurement: "Косинус фи",

    voltage: "Напряжение",
    amperage: "Сила тока",
    power: "Мощность",
    RPower: "Реактивная мощность",
    FPower: "Полная мощность",

    Ku: "Коэфицент усиления по напряжению ",
    Ki: "Коэфицент усиления по току",
    Kp: "Коэфицент усиления по мощности",
    "voltage max": "Максимальное напряжение",
    "voltage min": "Минимальное напряжение",

    //Трёхфазные
    "voltage 1": "Напряжение 1",
    "voltage 2": "Напряжение 2",
    "voltage 3": "Напряжение 3",
    "voltage 1_withUnitsOfMeasurement": "Напряжение 1 (Вольт)",
    "voltage 2_withUnitsOfMeasurement": "Напряжение 2 (Вольт)",
    "voltage 3_withUnitsOfMeasurement": "Напряжение 3 (Вольт)",

    "amperage 1": "Сила тока 1",
    "amperage 2": "Сила тока 2",
    "amperage 3": "Сила тока 3",
    "amperage 1_withUnitsOfMeasurement": "Сила тока 1 (Ампер)",
    "amperage 2_withUnitsOfMeasurement": "Сила тока 2 (Ампер)",
    "amperage 3_withUnitsOfMeasurement": "Сила тока 3 (Ампер)",

    "power 0": "Мощность 1",
    "power 1": "Мощность 2",
    "power 2": "Мощность 3",
    "power 0_withUnitsOfMeasurement": "Мощность 1 (Ватт)",
    "power 1_withUnitsOfMeasurement": "Мощность 2 (Ватт)",
    "power 2_withUnitsOfMeasurement": "Мощность 3 (Ватт)",
} as const;

type DNACodon = keyof typeof DNATranscriber;
type RNACodon = (typeof DNATranscriber)[DNACodon];

export default function toRna(
    dna: string,
    withUnitsOfMeasurement?: boolean
): string {
    console.log(dna);

    if (withUnitsOfMeasurement) {
        // if (isValidCodon(dna) ) {
        //     const dna_withUnitsOfMeasurement = dna+""
        //     return DNATranscriber[newRequest + "_withUnitsOfMeasurement"];
        // }

        const newRequest = dna + "_withUnitsOfMeasurement";
        if (isValidCodon(newRequest)) {
            return DNATranscriber[newRequest];
        }
    } else {
        if (isValidCodon(dna)) {
            return DNATranscriber[dna];
        }
    }

    return dna;

    // const codons = [...dna];
    // validateSequence(codons);

    // const transcribedRNA = codons.map((codon) => DNATranscriber[codon]);
    // return transcribedRNA;
}

function validateSequence(values: string[]): asserts values is DNACodon[] {
    if (!values.every(isValidCodon)) {
        throw Error("invalid sequence");
    }
}
function isValidCodon(value: string): value is DNACodon {
    return value in DNATranscriber;
}
