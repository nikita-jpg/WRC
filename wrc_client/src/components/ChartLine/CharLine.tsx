import { ArrowsAltOutlined } from "@ant-design/icons";
import { Button, DatePicker, DatePickerProps, Modal } from "antd";
import {
    CategoryScale,
    Chart as ChartJS,
    Legend,
    LineElement,
    LinearScale,
    PointElement,
    Title,
    Tooltip,
} from "chart.js";
import dayjs from "dayjs";
import { useState } from "react";
import { Line } from "react-chartjs-2";
import { Table } from "../Table/Table";
import { CharLineProps } from "./CharLineTypes";
import {
    filterDataByTime,
    getDataSets,
    getTableColumns,
    getTableRows,
    sortByDate,
} from "./utils";

import "./CharLine.scss";
import toRna from "./consts";

ChartJS.register(
    CategoryScale,
    LinearScale,
    PointElement,
    LineElement,
    Title,
    Tooltip,
    Legend
);

type Dataset = {
    label: string;
    data: number[];
    borderColor: string;
    backgroundColor: string;
};

export type TestProps = {
    labels: string[];
    dataset: Dataset[];
};

export const CharLine = (props: CharLineProps) => {
    const { data, className, title } = props;

    const options = {
        responsive: true,
        elements: {
            point: {
                pointStyle: "line",
            },
        },
        plugins: {
            legend: {
                position: "top" as const,
            },
            title: {
                display: true,
                text: toRna(title || "", true),
                //text: title,
            },
        },
    };

    let labelInfoData = [];
    let possibleDates = new Set<string>();

    for (const el of data) {
        for (const date of el.data) {
            possibleDates.add(date.dateTime.toDateString());
        }
    }

    const possibleDatesArray = Array.from(possibleDates);

    const [defaultDate] = useState(new Date(possibleDatesArray[0]));

    const [selectedDate, setSelectedDate] = useState(defaultDate);

    for (const el of data) {
        const sortedData = sortByDate(el.data);
        const selectedDates = filterDataByTime({
            requiredDate: selectedDate,
            info: sortedData,
        });

        labelInfoData.push({
            label: el.label,
            info: selectedDates,
            borderColor: el.borderColor,
            backgroundColor: el.backgroundColor,
        });
    }

    const datasets = getDataSets(labelInfoData);

    const [open, setOpen] = useState(false);

    const columns = getTableColumns(labelInfoData);
    const dataSource = getTableRows(datasets);

    const onChange: DatePickerProps["onChange"] = (date, dateString) =>
        setSelectedDate(new Date(dateString));

    return (
        <div className={className + " CharLine"}>
            <div className="header">
                <DatePicker
                    defaultValue={dayjs(defaultDate.toDateString())}
                    onChange={onChange}
                    disabledDate={(current) => {
                        return !possibleDatesArray.includes(
                            current.toDate().toDateString()
                        );
                    }}
                />

                <Button
                    icon={<ArrowsAltOutlined />}
                    onClick={() => setOpen(true)}
                />
            </div>
            <Table dataSource={dataSource} columns={columns} />
            <Modal
                // title={title || ""}
                centered
                open={open}
                onOk={() => setOpen(false)}
                onCancel={() => setOpen(false)}
                width={1000}
            >
                <Line options={options} data={datasets} />
            </Modal>
        </div>
    );
};
