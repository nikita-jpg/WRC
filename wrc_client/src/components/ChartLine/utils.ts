import { StringifyOptions } from "querystring";
import { CharLineDate } from "./CharLineTypes";

export const sortByDate = (data: CharLineDate[]) =>
    data.sort(
        (a: CharLineDate, b: CharLineDate) =>
            a.dateTime.getTime() - b.dateTime.getTime()
    );

export const getLabel = (data: CharLineDate) =>
    String(data.dateTime.toLocaleTimeString());

export const getLabels = (data: CharLineDate[]) =>
    data.map((data) => getLabel(data));

export const convertCharLineDateToOneDataset = (
    data: number[],
    label: string,
    borderColor?: string,
    backgroundColor?: string
) => ({
    label: label,
    data: data,
    borderColor: borderColor || "rgba(140, 0, 255, 0.5)",
    backgroundColor: backgroundColor || "rgba(140, 0, 255,0.5)",
});

export const getDataSet = (
    info: CharLineDate[],
    label = "value",
    borderColor?: string,
    backgroundColor?: string
) => {
    const data = info.map((inf) => inf.value);

    const datasets = convertCharLineDateToOneDataset(
        data,
        label,
        borderColor,
        backgroundColor
    );

    return datasets;
};

export type GetDataSets = {
    label: string;
    info: CharLineDate[];
    borderColor?: string;
    backgroundColor?: string;
};

export const getDataSets = (data: GetDataSets[]) => {
    const labels = getLabels(data[0].info);
    let datasets = [];

    for (let el of data) {
        datasets.push(
            getDataSet(el.info, el.label, el.borderColor, el.backgroundColor)
        );
    }

    return {
        labels,
        datasets,
    };
};

export const getTableRows = (data: {
    labels: string[];
    datasets: {
        label: string;
        data: number[];
        borderColor: string;
        backgroundColor: string;
    }[];
}) => {
    const getRow = (time: string, index: number) => {
        const keys = [...data.datasets.map((el) => el.label)] as const;

        const init = { key: String(index) } as Record<
            string,
            string | number
        > & {
            key: string;
        };

        init["time"] = time;
        keys.forEach((el, i) => {
            init[el] = data.datasets[i].data[index];
        });

        return init;
    };

    return data.labels.map((label, index) => getRow(label, index));
};

export const getTableColumns = (
    data: {
        label: string;
        data: number[];
        borderColor: string;
        backgroundColor: string;
    }[]
) =>
    data.map((el) => ({
        title: el.label,
        dataIndex: el.label,
        key: el.label,
    }));

type FilterDataByTime = {
    requiredDate?: Date;
    info: CharLineDate[];
};

export const filterDataByTime = ({ requiredDate, info }: FilterDataByTime) => {
    if (!requiredDate) {
        return info;
    }

    return info.filter(
        (el) => el.dateTime.toDateString() === requiredDate.toDateString()
    );

    // data.map((el) => ({
};
