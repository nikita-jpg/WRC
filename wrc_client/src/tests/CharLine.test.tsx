import "./matchMedia.mock";
import renderer from "react-test-renderer";
import { CharLine } from "../components/ChartLine/CharLine";

describe("<Toggle />", () => {
    it("should render OFF by default", () => {
        const tree = renderer
            .create(
                <CharLine
                    data={[
                        {
                            label: "voltage",
                            data: [
                                {
                                    dateTime: new Date(
                                        "2023-05-24T20:34:19.543Z"
                                    ),
                                    value: 100,
                                },
                            ],
                        },
                    ]}
                    className="page-dashboard__line"
                    title="voltage"
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();

        // render(
        //     <CharLine
        //         data={[
        //             {
        //                 label: "voltage",
        //                 data: [{ dateTime: new Date(), value: 1 }],
        //             },
        //         ]}
        //         className="page-dashboard__line"
        //         title="voltage"
        //     />
        // );

        // expect(screen.getByText(/OFF/)).toBeInTheDocument();
    });
});
