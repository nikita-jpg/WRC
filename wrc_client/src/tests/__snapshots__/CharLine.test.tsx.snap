// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toggle /> should render OFF by default 1`] = `
<div
  className="page-dashboard__line CharLine"
>
  <div
    className="header"
  >
    <div
      className="ant-picker css-dev-only-do-not-override-yp8pcc"
      onClick={[Function]}
    >
      <div
        className="ant-picker-input"
      >
        <input
          autoComplete="off"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          placeholder="Select date"
          readOnly={true}
          size={12}
          title=""
          value=""
        />
        <span
          className="ant-picker-suffix"
        >
          <span
            aria-label="calendar"
            className="anticon anticon-calendar"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="calendar"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
              />
            </svg>
          </span>
        </span>
        <span
          className="ant-picker-clear"
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          role="button"
        >
          <span
            aria-label="close-circle"
            className="anticon anticon-close-circle"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="close-circle"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
              />
            </svg>
          </span>
        </span>
      </div>
    </div>
    <button
      className="ant-btn css-dev-only-do-not-override-yp8pcc ant-btn-default ant-btn-icon-only"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <span
        aria-label="arrows-alt"
        className="anticon anticon-arrows-alt"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="arrows-alt"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M855 160.1l-189.2 23.5c-6.6.8-9.3 8.8-4.7 13.5l54.7 54.7-153.5 153.5a8.03 8.03 0 000 11.3l45.1 45.1c3.1 3.1 8.2 3.1 11.3 0l153.6-153.6 54.7 54.7a7.94 7.94 0 0013.5-4.7L863.9 169a7.9 7.9 0 00-8.9-8.9zM416.6 562.3a8.03 8.03 0 00-11.3 0L251.8 715.9l-54.7-54.7a7.94 7.94 0 00-13.5 4.7L160.1 855c-.6 5.2 3.7 9.5 8.9 8.9l189.2-23.5c6.6-.8 9.3-8.8 4.7-13.5l-54.7-54.7 153.6-153.6c3.1-3.1 3.1-8.2 0-11.3l-45.2-45z"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    className="ant-table-wrapper css-dev-only-do-not-override-yp8pcc"
  >
    <div
      className="ant-spin-nested-loading css-dev-only-do-not-override-yp8pcc"
    >
      <div
        className="ant-spin-container"
      >
        <div
          className="ant-table ant-table-small"
        >
          <div
            className="ant-table-container"
          >
            <div
              className="ant-table-content"
              onScroll={[Function]}
              style={{}}
            >
              <table
                style={
                  {
                    "tableLayout": "auto",
                  }
                }
              >
                <colgroup />
                <thead
                  className="ant-table-thead"
                >
                  <tr>
                    <th
                      className="ant-table-cell"
                      colSpan={null}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      rowSpan={null}
                      scope="col"
                      style={{}}
                    >
                      Время
                    </th>
                    <th
                      className="ant-table-cell"
                      colSpan={null}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      rowSpan={null}
                      scope="col"
                      style={{}}
                    >
                      voltage
                    </th>
                  </tr>
                </thead>
                <tbody
                  className="ant-table-tbody"
                >
                  <tr
                    className="ant-table-row ant-table-row-level-0"
                    data-row-key="0"
                    onClick={[Function]}
                    style={{}}
                  >
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      rowSpan={null}
                      style={{}}
                    >
                      8:34:19 PM
                    </td>
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      rowSpan={null}
                      style={{}}
                    >
                      1
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
        <ul
          className="ant-pagination ant-pagination-mini ant-table-pagination ant-table-pagination-right css-dev-only-do-not-override-yp8pcc"
          style={{}}
        >
          <li
            aria-disabled={true}
            className="ant-pagination-prev ant-pagination-disabled"
            onClick={[Function]}
            onKeyPress={[Function]}
            tabIndex={null}
            title="Previous Page"
          >
            <button
              className="ant-pagination-item-link"
              disabled={true}
              tabIndex={-1}
              type="button"
            >
              <span
                aria-label="left"
                className="anticon anticon-left"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="left"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                  />
                </svg>
              </span>
            </button>
          </li>
          <li
            className="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
            onClick={[Function]}
            onKeyPress={[Function]}
            tabIndex={0}
            title="1"
          >
            <a
              rel="nofollow"
            >
              1
            </a>
          </li>
          <li
            aria-disabled={true}
            className="ant-pagination-next ant-pagination-disabled"
            onClick={[Function]}
            onKeyPress={[Function]}
            tabIndex={null}
            title="Next Page"
          >
            <button
              className="ant-pagination-item-link"
              disabled={true}
              tabIndex={-1}
              type="button"
            >
              <span
                aria-label="right"
                className="anticon anticon-right"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="right"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                  />
                </svg>
              </span>
            </button>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
